consensus/ibft/hook/hook.go

```diff
@@ -0,0 +1,108 @@
+package hook
+
+import (
+	"github.com/0xPolygon/polygon-edge/state"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+// register multiple
+
+type ShouldWriteTransactionsFunc func(uint64) bool
+
+type ModifyHeaderFunc func(*types.Header, types.Address) error
+
+type VerifyHeaderFunc func(*types.Header) error
+
+type VerifyBlockFunc func(*types.Block) error
+
+type ProcessHeaderFunc func(*types.Header) error
+
+type PreCommitStateFunc func(*types.Header, *state.Transition) error
+
+type PostInsertBlockFunc func(*types.Block) error
+
+type Hooks interface {
+	ShouldWriteTransactions(uint64) bool
+	ModifyHeader(*types.Header, types.Address) error
+	VerifyHeader(*types.Header) error
+	VerifyBlock(*types.Block) error
+	ProcessHeader(*types.Header) error
+	PreCommitState(*types.Header, *state.Transition) error
+	PostInsertBlock(*types.Block) error
+}
+
+type HookManager struct {
+	ShouldWriteTransactionFunc ShouldWriteTransactionsFunc
+	ModifyHeaderFunc           ModifyHeaderFunc
+	VerifyHeaderFunc           VerifyHeaderFunc
+	VerifyBlockFunc            VerifyBlockFunc
+	ProcessHeaderFunc          ProcessHeaderFunc
+	PreCommitStateFunc         PreCommitStateFunc
+	PostInsertBlockFunc        PostInsertBlockFunc
+}
+
```

Comment:
(Same with `signer` package)

Why implement an interface in the same package. Consider removing the interface altogether and renaming `HookManager` to `Hooks` - its fields (callbacks) are exported anyway.

Reply:
I don't think interface must always be defined under the package where the module refer to the interface. It depends on the situation. Interface, Implementation, and the struct that refers to the Interface would be defined in separate package if many structs need to refer to the same interface. Ideally each structs to use the implementation must have its own interface. 

> It's just as easy given nothing more than...
This is not same as what I told. You need to define all fields and methods in this case. This way just makes writing mock flexible. This is the different topic.

Reply:
Well, there will be no big difference after I move the interface to the package. I will do that.

Reply:
Moved at ed475c40aa704cfb857acfb923963b13b664b8d9